api/coreservice_test.go

```diff
@@ -0,0 +1,149 @@
+	})
+	t.Run("invalid start and end height", func(t *testing.T) {
+		testData := &filterObject{FromBlock: "2", ToBlock: "1"}
+		filter, err := getTopicsAddress(testData.Address, testData.Topics)
+		require.NoError(err)
+		from, err := strconv.ParseUint(testData.FromBlock, 10, 64)
+		require.NoError(err)
+		to, err := strconv.ParseUint(testData.ToBlock, 10, 64)
+		require.NoError(err)
+
+		logs, err := svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter, nil, nil), from, to, uint64(0))
+		expectedErr := errors.New("invalid start and end height")
+		expectedValue := []*iotextypes.Log(nil)
+		require.Error(err)
+		require.Equal(expectedErr.Error(), err.Error())
+		require.Equal(expectedValue, logs)
+	})
+	t.Run("start block > tip height", func(t *testing.T) {
+		testData := &filterObject{FromBlock: "5", ToBlock: "5"}
+		filter, err := getTopicsAddress(testData.Address, testData.Topics)
+		require.NoError(err)
+		from, err := strconv.ParseUint(testData.FromBlock, 10, 64)
+		require.NoError(err)
+		to, err := strconv.ParseUint(testData.ToBlock, 10, 64)
+		require.NoError(err)
+
+		logs, err := svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filter, nil, nil), from, to, uint64(0))
+		expectedErr := errors.New("start block > tip height")
+		expectedValue := []*iotextypes.Log(nil)
+		require.Error(err)
+		require.Equal(expectedErr.Error(), err.Error())
+		require.Equal(expectedValue, logs)
+	})
+}
+
+func getTopicsAddress(addr []string, topics [][]string) (iotexapi.LogsFilter, error) {
+	var filter iotexapi.LogsFilter
+	for _, ethAddr := range addr {
+		ioAddr, err := ethAddrToIoAddr(ethAddr)
+		if err != nil {
+			return iotexapi.LogsFilter{}, err
+		}
+		filter.Address = append(filter.Address, ioAddr.String())
+	}
+	for _, tp := range topics {
+		var topic [][]byte
+		for _, str := range tp {
+			b, err := hexToBytes(str)
+			if err != nil {
+				return iotexapi.LogsFilter{}, err
+			}
+			topic = append(topic, b)
+		}
+		filter.Topics = append(filter.Topics, &iotexapi.Topics{Topic: topic})
+	}
+
+	return filter, nil
+}
+
+func BenchmarkLogsInRange(b *testing.B) {
+	svr, _, _, _, cleanCallback := setupTestServer()
+	defer cleanCallback()
+
+	filter := &filterObject{FromBlock: "0x1"}
+	from, _ := strconv.ParseInt(filter.FromBlock, 10, 64)
+	to, _ := strconv.ParseInt(filter.ToBlock, 10, 64)
+	var filters iotexapi.LogsFilter
+	for _, ethAddr := range filter.Address {
+		ioAddr, _ := ethAddrToIoAddr(ethAddr)
+		filters.Address = append(filters.Address, ioAddr.String())
+	}
+	for _, tp := range filter.Topics {
+		var topic [][]byte
+		for _, str := range tp {
+			b, _ := hexToBytes(str)
+			topic = append(topic, b)
+		}
+		filters.Topics = append(filters.Topics, &iotexapi.Topics{Topic: topic})
+	}
+
+	b.Run("five workers to extract logs", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			svr.web3Server.coreService.LogsInRange(logfilter.NewLogFilter(&filters, nil, nil), uint64(from), uint64(to), uint64(0))
```

Comment:
`FilterBlocksInRange` is the bottleneck of the entire benchmark. It would be better If it could be mocked here.

Reply:
But `mock api` needs create by these
```go
require := require.New(t)
ctrl := gomock.NewController(t)
defer ctrl.Finish()
core := mock_apicoreservice.NewMockCoreService(ctrl)
```

However, we can't use `t` inside `BenchmarkTest`, unless we don't use it.

Reply:
could you try 
```
	ctrl := gomock.NewController(b)
	defer ctrl.Finish()
	core := mock_apicoreservice.NewMockCoreService(ctrl)
```